In [14]:
import pandas as pd
from surprise import Dataset
from surprise import Reader
from surprise import KNNWithMeans
from surprise.model_selection import cross_validate
from surprise.model_selection import train_test_split
import numpy as np
import pickle
import random 
from surprise import accuracy

In [2]:
from surprise import NMF, SlopeOne, CoClustering, SVD, KNNWithMeans

In [3]:
random.seed(10)

In [4]:
ratings = pd.read_csv('ml-latest/ratings.csv')
print(ratings.head())
movies = pd.read_csv("movies_FINAL.csv")
print(movies.head())

   userId  movieId  rating   timestamp
0       1      307     3.5  1256677221
1       1      481     3.5  1256677456
2       1     1091     1.5  1256677471
3       1     1257     4.5  1256677460
4       1     1449     4.5  1256677264
   movieId                        title  year    rating  Western  Film-Noir  \
0        1                    Toy Story  1995  3.886649        0          0   
1        2                      Jumanji  1995  3.246583        0          0   
2        3             Grumpier Old Men  1995  3.173981        0          0   
3        4            Waiting to Exhale  1995  2.874540        0          0   
4        5  Father of the Bride Part II  1995  3.077291        0          0   

   IMAX  Musical  Sci-Fi  Adventure  ...  Action  War  Mystery  Animation  \
0     0        0       0          1  ...       0    0        0          1   
1     0        0       0          1  ...       0    0        0          0   
2     0        0       0          0  ...       0    0       

In [5]:
reader = Reader(rating_scale=(0.5, 5))

In [6]:
data = Dataset.load_from_df(ratings[["userId", "movieId", "rating"]], reader)

In [7]:
trainset, testset = train_test_split(data, test_size=.2)

In [8]:
#algo = SVD()

In [ ]:
filename = 'SVD_model.sav'

In [ ]:
algo = pickle.load(open(filename, 'rb'))

In [10]:
#algo.fit(trainset)

In [11]:
pickle.dump(algo, open(filename, 'wb'))

In [12]:
algo.predict(1, 307, r_ui =3.5,verbose=True)

user: 1          item: 307        r_ui = 3.50   est = 3.98   {'was_impossible': False}


Prediction(uid=1, iid=307, r_ui=3.5, est=3.976031439838299, details={'was_impossible': False})

In [16]:
predictions = algo.test(testset)
# Then compute RMSE

In [15]:
accuracy.rmse(predictions)

RMSE: 0.7950


0.7949583338171539

In [17]:
sim_options = {
    "name": "cosine",
    "user_based": False,
}

SVD - RMSE: 0.7950 - 910 MB
NMF - RMSE: 0.8777 - 910 MB
CoClustering - RMSE: 0.8759 - 824 MB

In [22]:
algo2 = NMF()
algo2.fit(trainset)

In [26]:
algo2.predict(1, 307, r_ui =3.5,verbose=True)

user: 1          item: 307        r_ui = 3.50   est = 4.14   {'was_impossible': False}


Prediction(uid=1, iid=307, r_ui=3.5, est=4.144688657770828, details={'was_impossible': False})

In [24]:
predictions2 = algo2.test(testset)
accuracy.rmse(predictions2)

RMSE: 0.8777


0.8776966157204822

In [35]:
filename2 = 'NMF_model.sav'

In [36]:
pickle.dump(algo, open(filename2, 'wb'))

In [32]:
d = data.build_full_trainset()

In [33]:
algo4 = CoClustering()
algo4.fit(d)

<ipython-input-33-b422e26ab8dc>:2: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  algo4.fit(d)


In [34]:
predictions4 = algo4.test(testset)
accuracy.rmse(predictions4)

RMSE: 0.8759


0.8759418420182284

In [37]:
filename3 = 'Cluster_model.sav'

In [38]:
pickle.dump(algo4, open(filename3, 'wb'))